### PR TITLE
Call walproposer from Rust

### PR DESCRIPTION
**Pull request by petuhovskiy**
Create Rust bindings for C functions from walproposer. This allows to write better tests with real walproposer code without spawning multiple processes and starting up the whole environment.

`make walproposer-lib` stage was added to build static libraries `libwalproposer.a`, `libpgport.a`, `libpgcommon.a`. These libraries can be statically linked to any executable to call walproposer functions.

`libs/walproposer/src/walproposer.rs` contains `test_simple_sync_safekeepers` to test that walproposer can be called from Rust to emulate sync_safekeepers logic. It can also be used as a usage example.

### CODE REVIEW

File: pgxn/neon/walproposer_compat.c [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+ *
+ * This is mostly for use in emitting messages.
+ */
+const char *
+timestamptz_to_str(TimestampTz t)
+{
+	static char buf[MAXDATELEN + 1];
+
+    snprintf(buf, sizeof(buf), "TimestampTz(%ld)", t);
+	return buf;
```
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply pgxn/neon/walproposer_compat.c:167]: I just realized that `static` arrays are ok in multi-process postgres usecase, but there can be issues if multiple threads are using this function at the same time.

File: libs/walproposer/src/walproposer.rs [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+            .unwrap()
+            .into_raw();
+
+        let mut safekeepers_list_vec = CString::new(config.safekeepers_list.join(","))
+            .unwrap()
+            .into_bytes_with_nul();
+        assert!(safekeepers_list_vec.len() == safekeepers_list_vec.capacity());
+        let safekeepers_list = safekeepers_list_vec.as_mut_ptr() as *mut i8;
+
+        let callback_data = Box::into_raw(Box::new(api)) as *mut ::std::os::raw::c_void;
```
Comment (arssher) [inline libs/walproposer/src/walproposer.rs:187]: Do we really need box of box here?

**Review by arssher — APPROVED** [review]
Impressive work!

File: pgxn/neon/walproposer.h [hunk 3]
```diff
@@ -666,7 +699,16 @@ extern WalProposer *WalProposerCreate(WalProposerConfig *config, walproposer_api
 extern void WalProposerStart(WalProposer *wp);
 extern void WalProposerBroadcast(WalProposer *wp, XLogRecPtr startpos, XLogRecPtr endpos);
 extern void WalProposerPoll(WalProposer *wp);
-extern void ParsePageserverFeedbackMessage(StringInfo reply_message,
-										   PageserverFeedback *rf);
+extern void WalProposerFree(WalProposer *wp);
+
+
+#define WPEVENT		1337	/* special log level for walproposer internal events */
```
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply pgxn/neon/walproposer.h:705]: Not used yet, I'm going to use it in the next patch.

File: libs/walproposer/build.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    // to bindgen, and lets you build up options for
+    // the resulting bindings.
+    let bindings = bindgen::Builder::default()
+        // The input header we would like to generate
+        // bindings for.
+        .header("bindgen_deps.h")
+        // Tell cargo to invalidate the built crate whenever any of the
+        // included header files changed.
+        .parse_callbacks(Box::new(CargoCallbacks))
+        .allowlist_type("WalProposer")
```
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply libs/walproposer/build.rs:71]: Not sure. Including everything will also capture internal postgres types and functions, i.e. everything reachable from includes. There is probably a way to include just a single file (`walproposer.h`), but explicit is better here IMO.

File: libs/walproposer/src/walproposer.rs (hunk 2 above)
Comment (petuhovskiy) [reply libs/walproposer/src/walproposer.rs:187]: `dyn ApiImpl` is a fat pointer type and `Box::into_raw(api)` cannot be casted to `void*`

File: pgxn/neon/walproposer.h [hunk 5]
```diff
@@ -590,6 +616,13 @@ typedef struct WalProposerConfig
 
 	/* Will be passed to safekeepers in greet request. */
 	uint64		systemId;
+
+	/* Will be passed to safekeepers in greet request. */
+	TimeLineID  pgTimeline;
+
+#ifdef WALPROPOSER_LIB
+	void *callback_data;
```
[the remark this answers is not in the document]
Comment (petuhovskiy) [reply pgxn/neon/walproposer.h:624]: Yes, I didn't think of that. Maybe I'll move it to walproposer_api in a later refactoring.

**Pull request merged**